Add sequential workshop tutorial format (pick-and-place seed) (#5104)
viamrobotics/docs: docs/tutorials/pick-and-place/_index.md

---
title: "Vision-Guided Pick-and-Place with the xArm6"
linkTitle: "Pick-and-Place Workshop"
type: "docs"
weight: 50
description: "Build a vision-guided robot that detects blocks by shape and sorts them into a bin with motion planning, from static arm poses to a Python perception loop."
authors: []
level: "Intermediate"
languages: ["python"]
viamresources:
  ["arm", "gripper", "camera", "vision", "motion", "frame_system", "switch"]
platformarea: ["mobility", "core"]
tags: ["tutorial", "workshop", "arm", "vision", "motion"]
workshop: "pick-and-place"
workshop_overview: true
time_estimate: "2 hours"
hardware:
  - "uFactory xArm6"
  - "Intel RealSense D435"
  - "uFactory finger gripper"
  - "System76 Meerkat"
companion_repo: "https://github.com/viam-devrel/pick-and-place"
no_list: true
---

> In this workshop you build a vision-guided robot that finds blocks by shape and picks each one into a bin. A `shape-detector` vision service locates blocks in camera space and feeds a `vision-segment` service (model `detections-to-segments`) that turns each detection into a point cloud segment the motion planner can grasp.

Reviewer: ```suggestion
In this workshop you build a vision-guided robot that finds blocks by shape and places each one into a bin. A `shape-detector` vision service locates blocks in camera space and feeds a `vision-segment` service (model `detections-to-segments`) that turns each detection into a point cloud segment the motion planner can grasp.
```

> The workshop is structured as six sequential phases, each ending with a working system state you can verify before moving on. The workshop has two milestones: by the end of Phase 4 you drive the robot from your own code through a static, pre-planned sequence (milestone one, a real and bankable win), and by the end of Phase 5 you close the loop with live perception so the robot detects, picks, and places blocks on its own (milestone two). At minimum, aim to complete the Phase 4 script.

Reviewer: ```suggestion
The workshop is structured as six sequential phases, each ending with a working system state you can verify before moving on. The workshop has two milestones: by the end of Phase 4 you drive the robot from your own code through a static, pre-planned sequence (milestone one), and by the end of Phase 5 you close the loop with live perception so the robot detects, picks, and places blocks on its own (milestone two). 
```

## What you'll build

<!-- ASSET P0 hero-hardware-overview (PHOTO): complete rig staged (xArm6 + gripper + wrist RealSense + blocks + bin + Meerkat). See plans/2026-07-02-pick-and-place-shot-list.md -->

> You will configure an xArm6 robotic arm fitted with a finger gripper and a wrist-mounted Intel RealSense depth camera. A shape-detection vision service finds blocks in camera space, and the Viam motion service plans and executes collision-free picks that place each block in the bin. By the end of Phase 5 you have a Python script you run from your laptop that drives the full detect-pick-place loop.

Reviewer: ```suggestion
You will configure an xArm6 robotic arm fitted with a two finger gripper and a wrist-mounted Intel RealSense depth camera. A shape-detection vision service finds blocks in camera space, and the Viam motion service plans and executes collision-free picks that place each block in the bin. By the end of Phase 5 you have a Python script you run from your laptop that drives the full detect-pick-place loop.
```

<!-- TODO: add hardware-overview photo to assets/tutorials/pick-and-place/ and reference via images frontmatter once available -->

> ## Hardware

Reviewer: ```suggestion
## Required Hardware
```

- **uFactory finger gripper**: the end-effector that grasps the blocks.
- **System76 Meerkat**: the on-robot mini-PC that runs the Viam machine server.

## Phases

Phases 1 through 5 are the core workshop. Phase 6 is optional.

1. **[Platform mental model](/tutorials/pick-and-place/platform-mental-model/)** (~15 min)
2. **[Configure resources and explore the app](/tutorials/pick-and-place/configure-resources/)** (~20 min)
3. **[Static positions and obstacles](/tutorials/pick-and-place/static-positions/)** (~20 min)
4. **[Control the robot from Python](/tutorials/pick-and-place/control-the-robot-from-python/)** (~15 min, milestone one)
5. **[Perception-guided picking](/tutorials/pick-and-place/perception-guided-picking/)** (~22 min, milestone two)
6. **[Inline module](/tutorials/pick-and-place/inline-module/)** (~20 min, optional)

## Prerequisites

This is a self-serve workshop, so confirm each of the following before you start:

- **Python 3.10 or newer.** Install it with [uv](https://docs.astral.sh/uv/getting-started/installation/) (recommended) or from [python.org](https://www.python.org/downloads/).

> - **The Viam Python SDK.** The companion `scripts/` project already declares `viam-sdk`, so `uv run` installs it for you in Phase 4. See the [Python SDK docs](https://python.viam.dev/) for reference. Pip works too if you prefer it.

Reviewer: scripts isn't listed until Companion Code, might be good to move that section before Prereqs, might be less confusing to just provide a link to install or a tabbed code snippet for UV, Pip, etc

- **A working terminal** on the machine you will run the Phase 4 and Phase 5 scripts from, typically your laptop rather than the robot's Meerkat.

> - **A Viam account with an accessible machine.** Log in at [app.viam.com](https://app.viam.com), open your machine, and confirm the green **Live** indicator before you begin.

Reviewer: Should we add something about accessing the machine with the resources in the machine config?

### Validate your environment

<!-- ASSET P1 env-validate (TERM): python3 --version and uv run --with viam-sdk import viam printing a version -->

Before starting Phase 4, confirm your environment is ready:

```sh
python3 --version   # 3.10 or newer
uv run --with viam-sdk python -c "import viam; print(viam.__version__)"   # prints a version
```

If either command fails, revisit the checklist above before continuing.

### Where to start

<!-- ASSET P1 live-indicator (UI+): machine page with the green Live indicator boxed -->

> - **Physical hardware ready:** start at [Phase 1](/tutorials/pick-and-place/platform-mental-model/).

Reviewer: ```suggestion
- **If the Physical hardware is up and running with Viam Server ** start at [Phase 1](/tutorials/pick-and-place/platform-mental-model/).
```

- **Provisioning your own hardware:** complete the hardware setup guide first (forthcoming), then return here for [Phase 1](/tutorials/pick-and-place/platform-mental-model/).

> Only the physical hardware, viam-agent, viam-server, and the frame calibration (the camera's mounting offset on the arm) may be pre-provisioned for you or come from the hardware setup guide. Configuring the arm, gripper, camera, and the vision and motion services is always your hands-on work in this workshop, starting in Phase 2.

Reviewer: ```suggestion

```

## Companion code

> All supporting files for this workshop live in the [viam-devrel/pick-and-place](https://github.com/viam-devrel/pick-and-place) repository.

Reviewer: Might be good to move this up and explain a bit that this is the reference

> - `config/` holds a machine config fragment and an obstacles template. Use them to check your work after you configure resources by hand, not as something to import wholesale.

Reviewer: ```suggestion
- `config/` holds a machine config fragment and an obstacles template. Use them to check your work after you configure resources by hand.
```

---
title: "Vision-Guided Pick-and-Place with the xArm6"
linkTitle: "Pick-and-Place Workshop"
type: "docs"
weight: 50

> description: "Build a vision-guided robot that detects blocks by shape and sorts them into a bin with motion planning, from static arm poses to a Python perception loop."

Reviewer: maybe reword from static arm poses to a Python perception loop. tutorial is cooler than this lets on

authors: []
level: "Intermediate"
languages: ["python"]
viamresources:
  ["arm", "gripper", "camera", "vision", "motion", "frame_system", "switch"]
platformarea: ["mobility", "core"]
tags: ["tutorial", "workshop", "arm", "vision", "motion"]
workshop: "pick-and-place"
workshop_overview: true
time_estimate: "2 hours"
hardware:
  - "uFactory xArm6"
  - "Intel RealSense D435"
  - "uFactory finger gripper"
  - "System76 Meerkat"
companion_repo: "https://github.com/viam-devrel/pick-and-place"
no_list: true
---

> In this workshop you build a vision-guided robot that finds blocks by shape and places each one into a bin. A `shape-detector` vision service locates blocks in camera space and feeds a `vision-segment` service (model `detections-to-segments`) that turns each detection into a point cloud segment the motion planner can grasp.

Reviewer: Might be too name heavy, might be good to soften?

You will use a camera to detect blocks, use those camera detections to find the corresponding area in the depth data, and segment the point cloud to tell the motion planner how to grasp the block.

Reviewer: this is a repeat of the text under the image, so maybe just delete this paragraph

2. **[Configure resources and explore the app](/tutorials/pick-and-place/configure-resources/)** (~15 min)
3. **[Static positions and obstacles](/tutorials/pick-and-place/static-positions/)** (~20 min)
4. **[Control the robot from Python](/tutorials/pick-and-place/control-the-robot-from-python/)** (~15 min, milestone one)
5. **[Perception-guided picking](/tutorials/pick-and-place/perception-guided-picking/)** (~22 min, milestone two)
6. **[Inline module](/tutorials/pick-and-place/inline-module/)** (~20 min, optional)

When you finish, the **[wrap-up](/tutorials/pick-and-place/wrap-up/)** reviews what you built and points to next steps.

## Companion code

All supporting files for this workshop live in the [viam-devrel/pick-and-place](https://github.com/viam-devrel/pick-and-place) repository. You do not need it to start, but the prerequisites below reference its `scripts/` project, and later phases pull from it.

- `config/` holds a machine config fragment and an obstacles template. Use them to check your work after you configure resources by hand.
- `scripts/` holds the starter script for Phase 4 and the reference solution for Phase 5.

## Prerequisites

> This is a self-serve workshop, so confirm each of the following before you start:

Reviewer: Might be good to slightly expand on "accessible machine
